src/main/java/net/patchworkmc/patcher/capabilityinject/CapabilityInjectRewriter.java

```diff
@@ -0,0 +1,162 @@
+		method.visitCode();
+
+		for (CapabilityInject inject : this.injects) {
+			String name = PREFIX + inject.getName();
+
+			// Put the event on the stack (1)
+			method.visitLdcInsn(inject.getType());
+			method.visitMethodInsn(Opcodes.INVOKESTATIC, "net/patchworkmc/api/capability/CapabilityRegisteredCallback", "event", "(Ljava/lang/Class;)Lnet/fabricmc/fabric/api/event/Event;", true);
+
+			// Put our handler on the stack (2)
+			Handle handle = new Handle(Opcodes.H_INVOKESTATIC, this.className, name, CAPABILITY_DESC, false);
+			method.visitInvokeDynamicInsn("onCapabilityRegistered", "()Lnet/patchworkmc/api/capability/CapabilityRegisteredCallback;", LambdaVisitors.METAFACTORY, Type.getMethodType(CAPABILITY_DESC), handle, Type.getMethodType(CAPABILITY_DESC));
+
+			// Register the event (0)
+			method.visitMethodInsn(Opcodes.INVOKEVIRTUAL, "net/fabricmc/fabric/api/event/Event", "register", "(Ljava/lang/Object;)V", false);
+		}
+
+		method.visitInsn(Opcodes.RETURN);
+		method.visitMaxs(2, 0);
+		method.visitEnd();
+	}
+
+	private void generateInjectHandlers() {
+		for (CapabilityInject inject : this.injects) {
+			String descriptor = "(Lnet/minecraftforge/common/capabilities/Capability;)V";
+			String name = PREFIX + inject.getName();
+			MethodVisitor handler = super.visitMethod(Opcodes.ACC_PUBLIC | Opcodes.ACC_STATIC | Opcodes.ACC_SYNTHETIC, name, descriptor, null, null);
+
+			if (handler != null) {
+				handler.visitCode();
+
+				// Get the capability (1)
+				handler.visitVarInsn(Opcodes.ALOAD, 0);
+
+				// Handle the capability (0)
+				if (inject.isMethod()) {
+					handler.visitMethodInsn(Opcodes.INVOKESTATIC, this.className, inject.getName(), "(Lnet/minecraftforge/common/capabilities/Capability;)V", false);
+				} else {
+					handler.visitFieldInsn(Opcodes.PUTSTATIC, this.className, inject.getName(), "Lnet/minecraftforge/common/capabilities/Capability;");
+				}
+
+				handler.visitInsn(Opcodes.RETURN);
+				handler.visitMaxs(1, 1);
+				handler.visitEnd();
+			}
+		}
+	}
+
+	public List<CapabilityInject> getInjects() {
+		return this.injects;
+	}
+
+	private class MethodScanner extends MethodVisitor {
+		private final int access;
+		private final String name;
+
+		MethodScanner(MethodVisitor parent, int access, String name, String descriptor) {
+			super(ASM7, parent);
+			this.access = access;
+			this.name = name;
+		}
+
+		public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
+			if (descriptor.equals("Lnet/minecraftforge/common/capabilities/CapabilityInject;")) {
+				if ((this.access & Opcodes.ACC_STATIC) == 0) {
+					Patchwork.LOGGER.error("Method " + this.name + " marked with an @CapabilityInject annotation was not static! All @CapabilityInject methods must be static.");
+					return null;
+				} else {
+					return new StringAnnotationHandler(
+							typeStr -> injects.add(new CapabilityInject(this.name, Type.getType(typeStr), true)));
```

Comment:
Instead of generating a shim caller method for already static methods, why not then just call the method directly? Then, shim methods would only be needed for fields.